modules/core/src/main/scala/org/tessellation/domain/statechannel/StateChannelValidator.scala

```diff
@@ -0,0 +1,63 @@
+package org.tessellation.domain.statechannel
+
+import cats.data.ValidatedNec
+import cats.effect.kernel.Async
+import cats.syntax.all._
+
+import org.tessellation.dag.snapshot.StateChannelSnapshotBinary
+import org.tessellation.domain.aci.StateChannelOutput
+import org.tessellation.domain.statechannel.StateChannelValidator.StateChannelValidationErrorOr
+import org.tessellation.ext.cats.syntax.validated._
+import org.tessellation.schema.address.Address
+import org.tessellation.security.SecurityProvider
+import org.tessellation.security.key.ops.PublicKeyOps
+import org.tessellation.security.signature.SignedValidator.SignedValidationError
+import org.tessellation.security.signature.{Signed, SignedValidator}
+
+import derevo.cats.{eqv, show}
+import derevo.derive
+
+trait StateChannelValidator[F[_]] {
+
+  def validate(stateChannelOutput: StateChannelOutput): F[StateChannelValidationErrorOr[StateChannelOutput]]
+
+}
+
+object StateChannelValidator {
+
+  def make[F[_]: Async: SecurityProvider](
+    signedValidator: SignedValidator[F]
+  ): StateChannelValidator[F] = new StateChannelValidator[F] {
+
+    override def validate(stateChannelOutput: StateChannelOutput): F[StateChannelValidationErrorOr[StateChannelOutput]] = for {
+      signaturesV <- signedValidator
+        .validateSignatures(stateChannelOutput.snapshot)
+        .map(_.errorMap[StateChannelValidationError](InvalidSigned))
+      addressSignatureV <- validateSourceAddressSignature(stateChannelOutput.address, stateChannelOutput.snapshot)
+    } yield signaturesV.productR(addressSignatureV).map(_ => stateChannelOutput)
+
+    private def validateSourceAddressSignature(
+      address: Address,
+      signedSC: Signed[StateChannelSnapshotBinary]
+    ): F[StateChannelValidationErrorOr[Signed[StateChannelSnapshotBinary]]] =
+      signedSC.proofs.existsM { proof =>
+        proof.id.hex.toPublicKey.map { signerSC =>
+          signerSC.toAddress =!= address
+        }
+      }.ifM(
+        NotSignedByAddressOwner
+          .asInstanceOf[StateChannelValidationError]
+          .invalidNec[Signed[StateChannelSnapshotBinary]]
+          .pure[F],
+        signedSC.validNec[StateChannelValidationError].pure[F]
+      )
+  }
+
+  @derive(eqv, show)
+  sealed trait StateChannelValidationError
+  case class InvalidSigned(error: SignedValidationError) extends StateChannelValidationError
```

Comment:
`InvalidSigned` -> `InvalidSignature`

Reply:
I have followed naming seen in `TransactionValidator` and `RumorValidator`.

Reply:
ok, anyway it's weird for me. @tbekas @TheMMaciek @kpudlik ?

Reply:
@marcinwadon it's not necessarily an invalid signature. It can be one of the following reasons:

```
InvalidSignatures
NotEnoughSignatures
DuplicateSigners
MissingSigners
```